### PR TITLE
[READY] Allows the supermatter's radiation collectors to produce research points in short bursts

### PR DESCRIPTION
**Pull request by deathride58**
Title. This allows the supermatter's radiation collectors to ~~mine bitcoin~~ produce research points by combusting tritium and oxygen via radiation. While a rad collector is producing research points, it is unable to produce power. Radiation collectors require a mix of tritium and oxygen to produce research points. These gasses are consumed at a very rapid pace when the rad collector is in research point production mode. Work together with science or atmos techs to increase the station's research point production at the temporary cost of power production!

This PR also moves the multitool's (buggy) ability to display how many watts a rad collector is producing to the rad collector's examine text, and moves tool interactions to the proper procs.

For reference, one room temperature tank with 50% tritium, 50% oxygen will produce a total of 50 research points with the most frequently used supermatter setup. More radiation will increase the rate at which research points are generated, and packing more moles of both gasses into the tank will increase the amount of time a rad collector will be able to generate research points.

:cl: deathride58
add: By using a multitool on a radiation collector, you can use it to generate research points! This requires a tank filled with a mix of tritium and oxygen.
tweak: The display of rad collectors' power production rates has been moved from a multitool interaction to examine text.
tweak: Radiation collectors now emit a ding when they run out of fuel.
/:cl:

I also plan on making the turbine capable of producing research points in the future, as it rarely ever sees use.

### CODE REVIEW

**Comment by Buggy123** [comment]
AT LAST

**Comment by nicbn** [comment]
Find a way to make players be aware of this ingame without needing to check changelog or the wiki.

**Comment by SgtxRootbeer** [comment]
@nicbn science powergamers will make sure engineering is aware of it.

**Comment by deathride58** [comment]
After doing some quick tests, I'd say this is safe to testmerge to get feedback from players, and figure out if it needs a bigger nerf or not.

For reference, a generic supermatter setup can produce a total of 10 points per minute for each radiation collector generating points.

A tank containing a perfect 50/50 mix of tritium and oxygen can operate for 5 minutes (producing 50 points total) at room temperature, and can function for 20 minutes (producing 200 points) when cooled to 73.15K. I might end up nerfing this by making radiation collectors a lot more gas hungry in research point generation mode, but I'd say it'd be best to see how it works out ingame before I nerf it, given the effort required to obtain tritium in the first place.

**Comment by Tlaltecuhtli** [comment]
200 points for 20 min of set up?

**Comment by factoryman942** [comment]
>10pts/collector/min
>6 collectors
>60pts/min with all 6 being used to make science, and no power production

doesn't science produce like 2442 pts/min at the moment or something?

**Comment by deathride58** [comment]
Science produced 1k points per minute before the recent buff. Now it produces 1.3k points per minute.

The low baseline regeneration rate is intended to give engineers a much better reason to run risky, high power production supermatter setups. The research point production rate scales linearly with the power that would have otherwise been produced by a rad collector. The low baseline regeneration rate helps prevent the mechanic from becoming a free, easy way to bypass the timelock that techwebs is supposed to be, and the scaling directly encourages experimentation.

Basically? Players should not be able to produce a major amount of research points without putting in some form of effort. There should also be an actual reason beyond memes and "hey what happens if i do this" to experiment with the supermatter.

**Comment by nicbn** [comment]
At least change the radiation collector descriptions to say it can produce power or research and modes can be changed with the mt.

**Comment by deathride58** [comment]
Already done. Examining rad collectors while they're off will show the current mode, its gas requirements, and a hint to use a multitool.

**Comment by epochayur** [comment]
Good PR. Even if it's not that many points, it's better than none, especially considering they're being generated by a department other than science, at the trade-off of other station resources (power) too. This is the exact sort of thing I wanna see happen to techweb

**Comment by nicbn** [comment]
Also why oxygen instead of just tritium?

**Comment by AdamElTablawy** [comment]
This is barely notable in a point difference.

**Comment by deathride58** [comment]
@AdamElTablawy As stated before, the amount of points generated per minute per point-producing rad collector scales with the amount of power that would have otherwise been produced. 10 points per minute is just the baseline, with the generic run-pure-n2-through-the-space-loop-and-filter-everything-out setup being used. Experimenting with it locally, I've been able to get a single rad collectors to peak at 60 points per minute, though the engine ended up delaminating before the tanks ran out. And that's exactly the kind of balance I was hoping to achieve, risks should have to be taken if the station wants to boost research point production.

Since there will be plenty of other ways to produce research points in the future, it'd be best to ensure that new research point production options are balanced with those future options in mind, otherwise we'll end up with something similar to the current state of chem balance.

@nicbn It's to help encourage cooperation with atmos techs or scientists, along with providing room for the explanation of "The collected radiation combusts the tritium to fuel the internal ~~bitcoin miner~~ research point generator, but this process requires oxygen."
It adds an additional step to getting an effective mix to produce research points with, so that an engineer can't simply set the supermatter to filter tritium, milk the power-producing rad collectors of tritium, and supercool the resulting tritium to get free research points all round with near-zero effort.

**Comment by epochayur** [comment]
unga bunga why point so low

**Comment by awesine** [comment]
solars is the supreme power generator

**Review by ninjanomnom — DISMISSED** [review]
Code seems fine but I'd feel hypocritical to merge something without testmerge if the author is asking for it given how much before I got angry about that.

**Comment by Tlaltecuhtli** [comment]
now rd produces almost 4000 points /min + pwr game drinks not sure if it will ever be used

File: code/modules/power/singularity/collector.dm [hunk 1]
```diff
@@ -27,23 +33,43 @@
 	. = ..()
 	AddComponent(/datum/component/rad_insulation, RAD_EXTREME_INSULATION, FALSE, FALSE)
 
+/obj/machinery/power/rad_collector/Initialize()
+	. = ..()
+	if(z in GLOB.station_z_levels)
+		linked_techweb = SSresearch.science_tech
```
Comment (Cyberboss) [inline code/modules/power/singularity/collector.dm]: What if it gets moved?

File: code/modules/power/singularity/collector.dm [hunk 2]
```diff
@@ -27,23 +33,43 @@
...
-			last_power-=power_produced
+		if(!bitcoinmining)
+			if(!loaded_tank.air_contents.gases[/datum/gas/plasma])
+				investigate_log("<font color='red'>out of fuel</font>.", INVESTIGATE_SINGULO)
+				playsound(src, 'sound/machines/ding.ogg', 50, 1)
+				eject()
+			else
+				loaded_tank.air_contents.gases[/datum/gas/plasma][MOLES] -= powerproduction_drain*drainratio
+				loaded_tank.air_contents.assert_gas(/datum/gas/tritium)
+				loaded_tank.air_contents.gases[/datum/gas/tritium][MOLES] += powerproduction_drain*drainratio
```
Comment (Cyberboss) [inline code/modules/power/singularity/collector.dm]: you do the same calculation twice

File: code/modules/power/singularity/collector.dm [hunk 3]
```diff
@@ -27,23 +33,43 @@
...
-		else
-			loaded_tank.air_contents.gases[/datum/gas/plasma][MOLES] -= 0.001*drainratio
-			loaded_tank.air_contents.assert_gas(/datum/gas/tritium)
-			loaded_tank.air_contents.gases[/datum/gas/tritium][MOLES] += 0.001*drainratio
-			loaded_tank.air_contents.garbage_collect()
-
-			var/power_produced = min(last_power, (last_power*RAD_COLLECTOR_STORED_OUT)+1000) //Produces at least 1000 watts if it has more than that stored
-			add_avail(power_produced)
-			last_power-=power_produced
+		if(!bitcoinmining)
```
Comment (Cyberboss) [inline code/modules/power/singularity/collector.dm]: Use early return with loaded_tank

File: code/modules/power/singularity/collector.dm [hunk 4]
```diff
@@ -27,23 +33,43 @@
...
+				last_power-=power_produced
+		else if(linked_techweb)
+			if(!loaded_tank.air_contents.gases[/datum/gas/tritium] || !loaded_tank.air_contents.gases[/datum/gas/oxygen])
+				playsound(src, 'sound/machines/ding.ogg', 50, 1)
+				eject()
+			else
+				loaded_tank.air_contents.gases[/datum/gas/tritium][MOLES] -= bitcoinproduction_drain*drainratio
+				loaded_tank.air_contents.gases[/datum/gas/oxygen][MOLES] -= bitcoinproduction_drain*drainratio
+				loaded_tank.air_contents.assert_gas(/datum/gas/carbon_dioxide)
+				loaded_tank.air_contents.gases[/datum/gas/carbon_dioxide][MOLES] += bitcoinproduction_drain*2*drainratio
```
Comment (Cyberboss) [inline code/modules/power/singularity/collector.dm]: 3 times here

File: code/modules/power/singularity/collector.dm [hunk 5]
```diff
@@ -121,6 +131,47 @@
...
+	default_unfasten_wrench(user, wrench, 0)
+	return TRUE
+
+/obj/machinery/power/rad_collector/crowbar_act(mob/living/user, obj/item/crowbar)
+	if(loaded_tank)
+		if(locked)
+			to_chat(user, "<span class='warning'>The controls are locked!</span>")
+			return TRUE
+		eject()
+		return TRUE
```
Comment (Cyberboss) [inline code/modules/power/singularity/collector.dm]: remove this line

File: code/modules/power/singularity/collector.dm [hunk 6]
```diff
@@ -121,6 +131,47 @@
...
+	else
+		to_chat(user, "<span class='warning'>There isn't a tank loaded!</span>")
+		return TRUE
```
Comment (Cyberboss) [inline code/modules/power/singularity/collector.dm]: and unindent this one

**Review by Cyberboss — DISMISSED** [review]

**Comment by optimumtact** [comment]
This need some kind of IC lore that makes sense.

**Comment by deathride58** [comment]
Lore can be thought of some time in the future.

Current explanation I've had in mind is that the collected radiation is used to combust the tritium + oxygen mix, which generates even more radiation to power a very experimental research point producing ASIC, which is manually connected to Nanotrasen's ~~mining pool~~ research point production farm. As ~~mining bitcoins~~ producing research points requires a *lot* of processing power for any sort of profit, this isn't nearly as profitable as the station's existing ~~mining farm~~ RnD servers, but it does help give the station a higher hashrate for its ~~bitcoin mining~~ research point production, which results in Nanotrasen giving the station a higher cut of ~~bitcoins~~ research points with each payout.

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Remove that one line please

File: code/modules/power/singularity/collector.dm (hunk 5 above)
Comment (deathride58) [reply code/modules/power/singularity/collector.dm]: Removing this line will make it so that ejecting a tank will result in the "There isn't a tank loaded!" message right after it's successfully ejected

**Review by Cyberboss — APPROVED** [review]

**Comment by BlueNothing** [comment]
So why is it that servers don't scale?

**Comment by deathride58** [comment]
@BlueNothing Balance, along with the lack of point generation for other departments.

Techwebs is meant to be act as a timegate for overpowered items. Allowing research point generation to scale just by replacing parts or breaching the server room to space just brings techwebs a step closer to old RnD.

**Comment by BlueNothing** [comment]
Except that things are already space-gated by seperate protolathes. To me,
techwebs (X)or seperated protolathes would be good, but both of them
together leaves RnD rather little to get out of it after mining gets their
stuff, and gives other departments stuff... if they can convince RnD to
give it to them instead of rushing mechs. The way it is now, other
departments are likely more underserved than they were before.

**Comment by optimumtact** [comment]
>space gated
you what

**Comment by BlueNothing** [comment]
I.E. no department can really claim supremacy like they used to through the
protolathes... except probably Robotics.

**Comment by BlueNothing** [comment]
Honestly, Robotics is still OP.
